File: Client_Root/Client/Assets/Scripts/Network/Messages/ToClient/WorldSnapShotToC.cs
using FlatBuffers;
using System.Collections.Generic;

public class WorldSnapShotToC : IMessage
{
    public const ushort MESSAGE_ID = MessageID.WorldSnapShotToC_ID;

    public int m_nTick;
    public float m_fTime;
    public List<FBS.EntityState> m_listEntityState = new List<FBS.EntityState>();

    public override ushort GetID()
    {
        return MESSAGE_ID;
    }

    public override IMessage Clone()
    {
        return null;
    }

    public override byte[] Serialize()
    {
        var OffsetEntityState = new Offset<FBS.EntityState>[m_listEntityState.Count];

        for (int i = 0; i < m_listEntityState.Count; ++i)
        {
            FBS.EntityState entityState = m_listEntityState[i];

            var pos = FBS.Data.Vector3.CreateVector3(m_Builder, entityState.Position.X, entityState.Position.Y, entityState.Position.Z);
            var rot = FBS.Data.Vector3.CreateVector3(m_Builder, entityState.Rotation.X, entityState.Rotation.Y, entityState.Rotation.Z);

            int[] keys = new int[entityState.BehaviorsMapKeyLength];
            float[] values = new float[entityState.BehaviorsMapValueLength];

            for (int j = 0; j < entityState.BehaviorsMapKeyLength; ++j)
            {
                keys[j] = entityState.GetBehaviorsMapKey(j);
                values[j] = entityState.GetBehaviorsMapValue(j);
            }

            var behaviorsMapKey = FBS.EntityState.CreateBehaviorsMapKeyVector(m_Builder, keys);
            var behaviorsMapValue = FBS.EntityState.CreateBehaviorsMapValueVector(m_Builder, values);

            FBS.EntityState.StartEntityState(m_Builder);
            FBS.EntityState.AddPlayerIndex(m_Builder, entityState.PlayerIndex);
            FBS.EntityState.AddPosition(m_Builder, pos);
            FBS.EntityState.AddRotation(m_Builder, rot);
            FBS.EntityState.AddBehaviorsMapKey(m_Builder, behaviorsMapKey);
            FBS.EntityState.AddBehaviorsMapValue(m_Builder, behaviorsMapValue);

            OffsetEntityState[i] = FBS.EntityState.EndEntityState(m_Builder);
        }

        var entityStates = FBS.WorldSnapShotToC.CreateEntityStatesVector(m_Builder, OffsetEntityState);

        FBS.WorldSnapShotToC.StartWorldSnapShotToC(m_Builder);
        FBS.WorldSnapShotToC.AddTick(m_Builder, m_nTick);
        FBS.WorldSnapShotToC.AddTime(m_Builder, m_fTime);
        FBS.WorldSnapShotToC.AddEntityStates(m_Builder, entityStates);
        var data = FBS.WorldSnapShotToC.EndWorldSnapShotToC(m_Builder);

        m_Builder.Finish(data.Value);

        return m_Builder.SizedByteArray();
    }

    public override bool Deserialize(byte[] bytes)
    {
        var buf = new ByteBuffer(bytes);

        var data = FBS.WorldSnapShotToC.GetRootAsWorldSnapShotToC(buf);

        m_nTick = data.Tick;
        m_fTime = data.Time;
        for (int i = 0; i < data.EntityStatesLength; ++i)
        {
            m_listEntityState.Add(data.GetEntityStates(i));
        }

        return true;
    }

    public override void OnReturned()
    {
        m_listEntityState.Clear();
    }
}

File: Server_Root/Server/RoomServer/Messages/WorldSnapShotToC.cpp
#include "stdafx.h"
#include "WorldSnapShotToC.h"
#include "WorldSnapShotToC_Data_generated.h"

WorldSnapShotToC::WorldSnapShotToC()
{
}

WorldSnapShotToC::~WorldSnapShotToC()
{
}

unsigned short WorldSnapShotToC::GetID()
{
	return MESSAGE_ID;
}

IMessage* WorldSnapShotToC::Clone()
{
	return NULL;
}

const char* WorldSnapShotToC::Serialize(int* pLength)
{
	//	최적화 고려.. 포인터 사용하는 방향.. (복사 속도때문에)

	list<Offset<EntityState>> listEntityState;
	for (map<int, IEntity*>::iterator it = m_mapEntity.begin(); it != m_mapEntity.end(); ++it)
	{
		int nPlayerIndex = it->first;
		IEntity* pEntity = it->second;

		FBSData::Vector3 pos(pEntity->GetPosition().x(), pEntity->GetPosition().y(), pEntity->GetPosition().z());
		FBSData::Vector3 rot(pEntity->GetRotation().x(), pEntity->GetRotation().y(), pEntity->GetRotation().z());

		vector<int> vecKeys;
		vector<float> vecValues;
		list<IBehavior*> listBehavior = pEntity->GetActivatedBehaviors();
		for (list<IBehavior*>::iterator it = listBehavior.begin(); it != listBehavior.end(); ++it)
		{
			vecKeys.push_back((*it)->GetMasterDataID());
			vecValues.push_back((*it)->GetTime());
		}

		auto BehaviorsMapKey = m_Builder.CreateVector(vecKeys);
		auto BehaviorsMapValue = m_Builder.CreateVector(vecValues);

		EntityStateBuilder entityState_builder(m_Builder);
		entityState_builder.add_PlayerIndex(nPlayerIndex);
		entityState_builder.add_Position(&pos);
		entityState_builder.add_Rotation(&rot);
		entityState_builder.add_BehaviorsMapKey(BehaviorsMapKey);
		entityState_builder.add_BehaviorsMapValue(BehaviorsMapValue);
		auto entityState = entityState_builder.Finish();

		listEntityState.push_back(entityState);
	}

	vector<Offset<EntityState>> vecOffsetEntityState;
	for (list<Offset<EntityState>>::iterator it = listEntityState.begin(); it != listEntityState.end(); ++it)
	{
		vecOffsetEntityState.push_back(*it);
	}
	auto entityStates = m_Builder.CreateVector(vecOffsetEntityState);

	WorldSnapShotToC_DataBuilder data_builder(m_Builder);
	data_builder.add_Tick(m_nTick);
	data_builder.add_Time(m_fTime);
	data_builder.add_EntityStates(entityStates);
	auto data = data_builder.Finish();

	m_Builder.Finish(data);

	*pLength = m_Builder.GetSize();

	return (char*)m_Builder.GetBufferPointer();
}

bool WorldSnapShotToC::Deserialize(const char* pChar)
{
	auto data = flatbuffers::GetRoot<WorldSnapShotToC_Data>((const void*)pChar);

	m_nTick = data->Tick();
	m_fTime = data->Time();

	//	Not necessary.. (Deserialize() is never called in server side..)
	for (int i = 0; i < data->EntityStates()->size(); ++i)
	{
		
	}

	return true;
}